### PR TITLE
Generate correct odata.nextlink when expanding multi-level contained entities

### PR DESCRIPTION
**Pull request by KenitoInc**
### Issues

*This pull request fixes issue #2262* 
*Additional work after PR #2287*
### Description

*Briefly describe the changes of this pull request.*
We have been generating an incorrect `navigationLink` when generating `nextLink` in multi-level expand in contained entities. This PR fixes that.

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 1]
```diff
@@ -95,6 +95,83 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
         }
 
+        /// <summary>
+        /// Generates a contained navigation link following the OData URL conventions for the entity represented by <paramref name="resourceContext"/> and the given
+        /// navigation property.
+        /// </summary>
+        /// <param name="resourceContext">The <see cref="ResourceContext"/> representing the entity for which the navigation link needs to be generated.</param>
+        /// <param name="navigationProperty">The EDM navigation property.</param>
+        /// <returns>The navigation link following the OData URL conventions.</returns>
+        public static Uri GenerateContainedNavigationPropertyLink(this ResourceContext resourceContext,
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:105]: >GenerateContainedNavigationPropertyLink [](start = 26, length = 39)

Can we use the above "GenerateNavigationPropertyLink" method with some changes?

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 2]
```diff
@@ -95,6 +95,83 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+            IEdmNavigationProperty navigationProperty)
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:106]: >ionProperty [](start = 42, length = 11)

the above method has "includeCast", shall we add such in this new method?

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 3]
```diff
@@ -95,6 +95,83 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+                // For Contained entity sets
+                if(containedEntitySet != null)
+                {
+                    NavigationPropertySegment navigationPropertySegment = new NavigationPropertySegment(containedEntitySet.NavigationProperty, containedEntitySet.ParentNavigationSource);
+                    internalOdataPath.Push(new KeySegment(ConventionsHelpers.GetEntityKey(currentLevelResourceContext), currentLevelResourceContext.StructuredType as IEdmEntityType, null));
+                    internalOdataPath.Push(navigationPropertySegment);
+                }
+
+                // For Non-Contained entity sets
+                if(entitySet != null)
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:155]: >if(entitySet != null) [](start = 16, length = 21)

else?

Comment (xuzhg) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:155]: >if(entitySet != null) [](start = 16, length = 21)

how about if the path starts with singleton? the navigation source is "IEdmSingleton"

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 4]
```diff
@@ -95,6 +95,83 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+            // Loop through the all ExpandedResource(s).
+            // We generate OData Path Segments for all levels.
+            while (currentLevelResourceContext != null)
+            {
+                IEdmNavigationSource navigationSource = currentLevelResourceContext.NavigationSource;
+                IEdmEntitySet entitySet = navigationSource as IEdmEntitySet;
+                IEdmContainedEntitySet containedEntitySet = navigationSource as IEdmContainedEntitySet;
+
+                // For Contained entity sets
+                if(containedEntitySet != null)
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:147]: >if(containedEntitySet != null) [](start = 16, length = 30)

It could have the type cast in the path? for example:

~/customers/NS.VipCustomer/ContainedOrders/NS.VipOrder

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs (hunk 1 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:106]: Added `includeCast`
#Resolved

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs (hunk 2 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs]: Added `includeCast`
#Resolved

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs (hunk 3 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:162]: Added capabilities for Singleton
#Resolved

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs (hunk 4 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:154]: Added type cast.
#Resolved

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs (hunk 3 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:162]: We don't need else here.
We are checking if any of the following is not null and adding the appropriate Path Segments
- entitySet
- containedEntitySet
- singleton

Comment (habbes) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:162]: Also, is it possible for there to be a valid segment other than the 3 you have tested for? Like for example a property segment?

File: src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSetSerializer.cs [hunk 5]
```diff
@@ -486,10 +486,7 @@ private static Uri GetNestedNextPageLink(ODataSerializerContext writeContext, in
             if (writeContext.NavigationSource.NavigationSourceKind() == EdmNavigationSourceKind.ContainedEntitySet)
             {
                 // Contained navigation.
-                Uri idlink = linkBuilder.BuildIdLink(writeContext.ExpandedResource);
-
-                var link = idlink.ToString() + "/" + writeContext.NavigationProperty.Name;
-                navigationLink = new Uri(link);
+                navigationLink = writeContext.ExpandedResource.GenerateContainedNavigationPropertyLink(writeContext.NavigationProperty, false);
```
Comment (habbes) [inline src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSetSerializer.cs:489]: Maybe add `includeCast: false` for clarity?

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs [hunk 6]
```diff
@@ -54,6 +54,24 @@ public void GenerateNavigationLink_WorksToGenerateExpectedNavigationLink_ForEnti
             Assert.Equal(expectedNavigationLink, uri.AbsoluteUri);
         }
 
+        [Theory]
+        [InlineData(false, "http://localhost/MyOrders(42)/OrderLines")]
+        [InlineData(true, "http://localhost/MyOrders(42)/NS.MyOrder/OrderLines")]
```
Comment (habbes) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs:59]: Could you also add a test where the last path segment is a type cast?

**Review by habbes — CHANGES_REQUESTED** [review]

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 7]
```diff
@@ -95,6 +95,102 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+                // For Singletons
+                if(singleton != null)
+                {
+                    SingletonSegment singletonSegment = new SingletonSegment(singleton);
+                    internalOdataPath.Push(singletonSegment);
+                }
+
+                currentLevelResourceContext = currentLevelResourceContext.SerializerContext.ExpandedResource;
+
+                if (includeCast & currentLevelResourceContext != null)
```
[the remark this answers is not in the document]
Comment (habbes) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:178]: Just checked the docs and realized that `&` is also used as an AND operator that does not short-circuit when the first operand is false. But in this case, is it necessary to evaluate `currentLevelResourceContext != null` if `includeCast` is false?

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 8]
```diff
@@ -95,6 +95,102 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+                }
+
+                // For Singletons
+                if(singleton != null)
+                {
+                    SingletonSegment singletonSegment = new SingletonSegment(singleton);
+                    internalOdataPath.Push(singletonSegment);
+                }
+
+                currentLevelResourceContext = currentLevelResourceContext.SerializerContext.ExpandedResource;
```
[the remark this answers is not in the document]
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:176]: Actually, it's the reverse. In `Employee`, the expanded resource is `Store`.

File: src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSetSerializer.cs (hunk 5 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSetSerializer.cs:489]: I have updated to ` /*includeCast*/ false`

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs (hunk 3 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:162]: 1. I have update the code to use pattern matching
2. I don't think there are other valid segments in this use case. I will be interested in getting suggestions.

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 9]
```diff
@@ -95,6 +95,102 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+                    internalOdataPath.Push(new TypeSegment(currentLevelResourceContext.StructuredType, navigationSource: null));
+                }
+            }
+
+            // We have added the path segments from the last to the first
+            // We then need to order them from the first to the last.
+            foreach (var i in internalOdataPath)
+            {
+                odataPathSegments.Add(i);
+            }
```
[the remark this answers is not in the document]
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:189]: #Resolved

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs (hunk 6 above)
Comment (KenitoInc) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs:59]: Actually in this test the last segment has a typecast

**Review by habbes — APPROVED** [review]

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 10]
```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+        /// <param name="includeCast">Represents whether the generated link should have a cast segment representing a type cast.</param>
+        /// <returns>The navigation link following the OData URL conventions.</returns>
+        public static Uri GenerateContainedNavigationPropertyLink(this ResourceContext resourceContext,
+            IEdmNavigationProperty navigationProperty, bool includeCast)
+        {
+            if (resourceContext == null)
+            {
+                throw Error.ArgumentNull("resourceContext");
+            }
+            if (resourceContext.InternalUrlHelper == null)
```
[the remark this answers is not in the document]
Comment (mikepizzo) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:113]: Yes; the empty line should be there per style guidelines, and will be required once we enforce style guidelines.

---
In reply to: [572719349](https://github.com/OData/WebApi/pull/2400#discussion_r572719349) [](ancestors = 572719349)

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 11]
```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+
+            navigationPathSegments.Add(new NavigationPropertySegment(navigationProperty, navigationSource: null));
+
+            string link = resourceContext.InternalUrlHelper.CreateODataLink(navigationPathSegments);
+            if (link == null)
+            {
+                return null;
+            }
+
+            return new Uri(link);
```
Comment (mikepizzo) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:133]: >link [](start = 27, length = 4)

will link always be an absolute Uri?

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 12]
```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+                    reversedOdataPathSegments.Push(new KeySegment(ConventionsHelpers.GetEntityKey(currentLevelResourceContext), currentLevelResourceContext.StructuredType as IEdmEntityType, null));
+                    reversedOdataPathSegments.Push(entitySetSegment);
+                }
+
+                // For Singletons
+                if(navigationSource is IEdmSingleton singleton)
+                {
+                    SingletonSegment singletonSegment = new SingletonSegment(singleton);
+                    reversedOdataPathSegments.Push(singletonSegment);
+                }
```
Comment (mikepizzo) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:171]: if we encounter a navigation source that is a singleton, or non-contained entity set (and we have a key), shouldn't we reset the reversedOdatapathSegments in order to return the canonical navigation path?  i.e., if we have an indirect path like users(user1)/groups(group1)/members(user5)/address, where address is a contained navigation property and members is a navigation path to a user (bound to a non-contained entity set "users") then shouldn't we ignore the users(user1)/groups(group1) and just return users(user5)/address?  And similarly, if company is a top level singleton, shouldn't users(user1)/bestFriend/friends(user7)/company/address just return company/address?

What happens if we have groups(group1)/owner/address (where "owner is a single-value non-contained nav prop on a group), would we have to return the (non-canonical) groups(group1)/owner/address, or would we have the id of the owner so that we could return users(userId)/address?

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs [hunk 13]
```diff
@@ -54,6 +54,24 @@ public void GenerateNavigationLink_WorksToGenerateExpectedNavigationLink_ForEnti
             Assert.Equal(expectedNavigationLink, uri.AbsoluteUri);
         }
 
+        [Theory]
+        [InlineData(false, "http://localhost/MyOrders(42)/OrderLines")]
```
Comment (mikepizzo) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs:58]: >http://localhost/MyOrders(42)/OrderLines [](start = 28, length = 40)

what if you have a single-value, non-contained, navigation source in the path.  i.e., assuming a Customer has a non-contained single-valued navigation property "mostRecentOrder" and the request was:
[http://localhost/Customers(42)/mostRecentOrder/OrderLines](http://localhost/Customers(42)/mostRecentOrder/OrderLines)

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 14]
```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+        /// <summary>
+        /// Generates a contained navigation link following the OData URL conventions for the entity represented by <paramref name="resourceContext"/> and the given
+        /// navigation property.
+        /// </summary>
+        /// <param name="resourceContext">The <see cref="ResourceContext"/> representing the entity for which the navigation link needs to be generated.</param>
+        /// <param name="navigationProperty">The EDM navigation property.</param>
+        /// <param name="includeCast">Represents whether the generated link should have a cast segment representing a type cast.</param>
+        /// <returns>The navigation link following the OData URL conventions.</returns>
+        public static Uri GenerateContainedNavigationPropertyLink(this ResourceContext resourceContext,
+            IEdmNavigationProperty navigationProperty, bool includeCast)
```
Comment (mikepizzo) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:107]: >includeCast [](start = 60, length = 11)

how does the caller know whether to includeCast? Shouldn't we determine this based on whether the navigationProperty is defined on the same type as the navigationSource?  If so, we don't need the cast, if not we have to include it.

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 15]
```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+                // For Singletons
+                if(navigationSource is IEdmSingleton singleton)
+                {
+                    SingletonSegment singletonSegment = new SingletonSegment(singleton);
+                    reversedOdataPathSegments.Push(singletonSegment);
+                }
+
+                currentLevelResourceContext = currentLevelResourceContext.SerializerContext.ExpandedResource;
+
+                if (includeCast && currentLevelResourceContext != null)
```
Comment (mikepizzo) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:175]: >includeCast [](start = 20, length = 11)

we should determine this for each step depending on whether the type that the segment we are appending is defined on matches the type of the previous segment.

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 16]
```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+            IList<ODataPathSegment> odataPathSegments = new List<ODataPathSegment>();
+            Stack<ODataPathSegment> reversedOdataPathSegments = new Stack<ODataPathSegment>();
+
+            ResourceContext currentLevelResourceContext = resourceContext;
+
+            // Loop through the all ExpandedResource(s).
+            // We generate OData Path Segments for all levels.
+            while (currentLevelResourceContext != null)
+            {
+                IEdmNavigationSource navigationSource = currentLevelResourceContext.NavigationSource;
```
Comment (mikepizzo) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:148]: >NavigationSource [](start = 84, length = 16)

is there ever a case where navigationSource is null? BadEntitySet? Are those cases properly handled?

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 17]
```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+                // For Contained entity sets
+                if (navigationSource is IEdmContainedEntitySet containedEntitySet)
+                {
+                    NavigationPropertySegment navigationPropertySegment = new NavigationPropertySegment(containedEntitySet.NavigationProperty, containedEntitySet.ParentNavigationSource);
+                    reversedOdataPathSegments.Push(new KeySegment(ConventionsHelpers.GetEntityKey(currentLevelResourceContext), currentLevelResourceContext.StructuredType as IEdmEntityType, null));
+                    reversedOdataPathSegments.Push(navigationPropertySegment);
+                }
+
+                // For Non-Contained entity sets
+                if(navigationSource is IEdmEntitySet entitySet)
```
Comment (mikepizzo) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:159]: What happens if we have groups(group1)/owner/address (where "owner is a single-value non-contained nav prop on a group), would we have to return the (non-canonical) groups(group1)/owner/address, or would rely on being able to get the id of the user from ResourceContext so that we could return users(userId)/address?

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs [hunk 18]
```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
...
+                {
+                    EntitySetSegment entitySetSegment = new EntitySetSegment(entitySet);
+                    reversedOdataPathSegments.Push(new KeySegment(ConventionsHelpers.GetEntityKey(currentLevelResourceContext), currentLevelResourceContext.StructuredType as IEdmEntityType, null));
```
Comment (mikepizzo) [inline src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:162]: >GetEntityKey [](start = 85, length = 12)

is there ever a case where we won't be able to get the entity key in the resource context (i.e., if the key properties weren't selected)? How do we handle such a case? Or do we enforce that the key properties are always selected in this case?

**Review by mikepizzo — CHANGES_REQUESTED** [review]
:clock1:

**Comment by mikepizzo** [comment]
Bunch of questions about how this works for edge cases. Perhaps those edge cases never apply due to how this is called, but want to be sure we think through and make sure they are not applicable and that we have relevant assertions in the code.

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs [hunk 19]
```diff
@@ -89,6 +107,24 @@ public void GenerateNavigationLink_WorksToGenerateExpectedNavigationLink_ForSing
             Assert.Equal(expectedNavigationLink, uri.AbsoluteUri);
         }
 
+        [Theory]
```
Comment (mikepizzo) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/LinkGenerationHelpersTest.cs:110]: >Theory [](start = 9, length = 6)

also add a test case for a request that includes a function in the path.

File: src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs (hunk 11 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs:133]: `CreateODataLink` will always generate an absolute uri.
